packages/ckeditor5-ui/tests/toolbar/toolbarview.js

```diff
@@ -447,6 +447,47 @@ describe( 'ToolbarView', () => {
 			expect( items.get( 4 ).name ).to.equal( 'foo' );
 		} );
 
+		it( 'accepts configuration object', () => {
+			view.fillFromConfig( { items: [ 'foo', 'bar', 'foo' ] }, factory );
+
+			const items = view.items;
+			expect( items ).to.have.length( 3 );
+			expect( items.get( 0 ).name ).to.equal( 'foo' );
+			expect( items.get( 1 ).name ).to.equal( 'bar' );
+			expect( items.get( 2 ).name ).to.equal( 'foo' );
+		} );
+
+		it( 'removes items listed in `removeItems`', () => {
+			view.fillFromConfig(
+				{
+					items: [ 'foo', 'bar', 'foo' ],
+					removeItems: [ 'foo' ]
+				},
+				factory
+			);
+
+			const items = view.items;
+			expect( items ).to.have.length( 1 );
+			expect( items.get( 0 ).name ).to.equal( 'bar' );
+		} );
+
+		it( 'deduplicates consecutive separators after removing items listed in `removeItems`', () => {
```

Comment:
Funny, because if for any weird reason someone had two separators already, we're changing the behavior for them. I'd add it to the merge commit message as "MINOR BREAKING CHANGE".

cc @oleq We're changing this because the `config.toolbar.removeItems` option might lead to having two separators one after another if you removed all buttons between them. Do you think it's fine if `fillFromConfig()` makes such an optimization? The other option could be to extract that logic to `normalizeToolbarConfig()`. It'd actually might make more sense there, taken the "normalization" aspect. WDYT, guys?

Reply:
On one hand, I am happy to change whatever you all think is the right thing to do. Still learning habits of this tribe. On the other hand, the name suggests it's about normalization, not sanitization/preprocessing.

Reply:
Or, perhaps, take out the entire logic of handling the config to a separate class. It's not ToolbarViews responsibility to deal with all the config logic. Especially now, when it grew in complexity.

Reply:
Entire logic has stayed in the `fillFromConfig()`. We do a few steps of filtering to cover all the cases.

Reply:
I'm OK with this optimization. Although I need to say I cannot predict to what extent this is a breaking change for integrators. I guess this could fail if someone relies on the order of buttons when they access them via `editor.ui.view.toolbar.items.get( index )` for some reason.

Also FYI, there are already some micro-optimizations in CSS when it comes to grouping separators in the "kebab" drop-down, for instance, if the first grouped item is a separator it is not rendered in the drop-down (because the toolbar should not start with a separator). Maybe this is the right way to go instead of changing the JS logic? This would be way cheaper and safer I guess. `.ck-toolbar__separator + .ck-toolbar__separator { display: none }`? 

P.S. Since we optimize separators, let's make sure line breaks are also optimized to avoid scenarios like `[ item, '-', '-', item ]`.

Reply:
Yup! Can confirm that \`.ck-toolbar\_\_separator + .ck-toolbar\_\_separator { display: none }\` works well.

Reply:
To me, it makes more sense to filter this in JS whenever we can. No one actually wants two separators one by one. When combinging `toolbar.items` with `toolbar.removeItems`, removing also duplicated and trailed separators seems to be a reasonable optimization.

As for whether this is a breaking change – certainly. But not a major one. I think it's acceptable.

> P.S. Since we optimize separators, let's make sure line breaks are also optimized to avoid scenarios like `[ item, '-', '-', item ]`.

Yeah, makes sense too. Good catch.

Reply:
Fixed leading and trailing cases for both of the separators along with duplicates.